File: ports/java/src/network/socket/BcsTcpSocket.java
package bcs.network.socket;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

// BcsTcpSocket abstracts a network client connection for Java
public class BcsTcpSocket {
    private Socket socket;
    private PrintWriter out;
    private BufferedReader in;

    public void connectToHost(String host, int port) throws IOException {
        socket = new Socket(host, port);
        out = new PrintWriter(socket.getOutputStream(), true);
        in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
    }

    public void write(String data) {
        if (out != null) {
            out.print(data);
            out.flush();
        }
    }

    public String readLine() throws IOException {
        if (in != null) {
            return in.readLine();
        }
        return null;
    }

    public void disconnect() throws IOException {
        if (socket != null && !socket.isClosed()) {
            socket.close();
        }
    }
}

File: ports/csharp/src/network/socket/BcsTcpSocket.cs
using System;
using System.IO;
using System.Net.Sockets;
using System.Threading.Tasks;

namespace BCS.Network.Socket {
    // BcsTcpSocket abstracts a network client connection for C#
    public class BcsTcpSocket {
        private TcpClient _client;
        private NetworkStream _stream;
        private StreamReader _reader;
        private StreamWriter _writer;

        public BcsTcpSocket() {
            _client = new TcpClient();
        }

        public async Task ConnectToHostAsync(string host, int port) {
            await _client.ConnectAsync(host, port);
            _stream = _client.GetStream();
            _reader = new StreamReader(_stream);
            _writer = new StreamWriter(_stream) { AutoFlush = true };
        }

        public async Task WriteAsync(string data) {
            await _writer.WriteAsync(data);
        }

        public async Task<string> ReadLineAsync() {
            return await _reader.ReadLineAsync();
        }

        public void Disconnect() {
            _stream?.Close();
            _client?.Close();
        }
    }
}
